### DISCUSSION TITLE
linking and loading (macOS) documentation

### DISCUSSION THREAD

**Discussion by nyholku** [bytedeco/javacpp#534] [Q&A] [answered]
Hi

I'm tipping my tow into the water and trying to see if I could make a start of calling some OpenCascade (OCC) methods from Java using JavaCPP.

I've managed to scrape together something that compiles but does not load (UnsatisfiedLinkError).

I have hard time finding a description of how to the linking and loading is controlled in by/in/with JavaCPP.

Is that documented somewhere?

Specifically:

OCC has thousands of header files, on my machine at `/usr/local/include/opencascade` and tens of dylibs at `/usr/local/lib`.

How do I tell JavaCPP about these libraries and how do I load them in Java at runtime?

For trivial examples it seems to work by magic but for anything more complex I stumped.

Looking further down the road I would not like to link to libaries at `/usr/local/lib`, but rather include them in a macOS bundle along the rest of the application for best end user experience.

cheers Kusti

**Comment by saudet** [comment]
Please start here and let me know if you have any questions: https://github.com/bytedeco/javacpp/wiki/Mapping-Recipes

**Comment by nyholku** [comment, marked as answer]
Thanks I've read that but I find that a bit implicit and with other hurdles a noob faces it is difficult to know if one is on the right track or not.

For example I created the following (see end of post).

I pressumed that 

```
        linkpath = {"/usr/local/lib"},//
        link = {"TKernel"}, //
```
 this would link my library to refer to  `/usr/local/lib/libTKernel.dylib`, is that correct?

Is there anything else I need to do to make my code library load the native libraries?

Unfortunately I cannot test this as I've yet not found a simple enough header in OpenCascade that would not make JavaCPP choke and which would contain a simple function I can trivially call. 



```

@Properties(
    value = @Platform(
        includepath = {"/usr/local/include/opencascade"},//
        preloadpath = {""},//
        linkpath = {"/usr/local/lib"},//
        link = {"TKernel"}, //
        include = {"Standard_Handle.hxx","FSD_Base64Decoder.hxx"},//
        preload = {""}//
    ),
    target = "TKernel"
)
public class TKernelConfig implements InfoMapper {

    public void map(InfoMap infoMap) {
    	infoMap.put(new Info("Standard_EXPORT", "").cppTypes().annotations());
    	infoMap.put(new Info("Standard_Real", "double").cppTypes().annotations());
    	infoMap.put(new Info("DEFINE_STANDARD_ALLOC", "").cppTypes().annotations());
    	infoMap.put(new Info("TARGET_OS_IPHONE","0").cppTypes().annotations());

    }
}
```

**Reply by nyholku** [reply]
Related issue, how do I tell the clan to use '-std=c++11' when compiling?

**Reply by saudet** [reply]
> this would link my library to refer to  `/usr/local/lib/libTKernel.dylib`, is that correct?

Yes, that's how it works.

> Is there anything else I need to do to make my code library load the native libraries?

The error message should tell you what's missing, so you can add it. That has nothing to do with JavaCPP.

> Unfortunately I cannot test this as I've yet not found a simple enough header in OpenCascade that would not make JavaCPP choke and which would contain a simple function I can trivially call.

You'll need to find the "root" header file and start from there, yes.

> Related issue, how do I tell the clan to use '-std=c++11' when compiling?

Something like this: https://github.com/bytedeco/javacpp-presets/blob/master/tensorflow-lite/src/main/java/org/bytedeco/tensorflowlite/presets/tensorflowlite.java#L44

**Reply by nyholku** [reply]
Thanks, got that now. :)
